### PR TITLE
[YUNIKORN-204] Extended UpdateResponse with UpdatedApplications

### PR DESCRIPTION
**Pull request by kingamarton**

### CODE REVIEW

File: scheduler-interface-spec.md [hunk 1]
```diff
@@ -270,11 +270,21 @@ message UpdateResponse {
...
   // Accepted Node Registrations
-  repeated AcceptedNode acceptedNodes = 9;
+  repeated AcceptedNode acceptedNodes = 10;
+}
+
+message UpdatedApplication {
+  // The application ID that was updated
+  string applicationID = 1;
+  // State of the application
+  string state = 2;
```
Comment (yangwwei) [inline scheduler-interface-spec.md:287]: please add 2 more fields:
1. stateTransitionTimestamp
2. a string message that can help to explain what happened

Comment (kingamarton) [reply scheduler-interface-spec.md:287]: done

**Comment by wilfred-s** [comment]
I do have a problem here. 

Application updates are not linked to state changes. The only update of an application that triggers a state change is adding the first request to allocate a resource, i.e. the first ask will move from a `new` state to an `accepted` state.
All other state changes are based on scheduler actions which are not in response to an application update. Adding state changes to the `UpdateResponse` is thus not the correct thing to do. First of all: there is no update to respond to. It is a self contained change by the scheduler.
Second point is that there is a large amount of overhead for this message.

We need to roll this message into the one set of updates.
If we add a status update in the way described we have 3 ways of communicating these updates: `OutstandingResourceRequest`, `UpdateContainerSchedulingStateRequest` and now `UpdateResponse`.

This is two too many.

**Comment by yangwwei** [comment]
Hi @wilfred-s  Not sure I completely understand the point.
As we have the app state maintained in `pkg/cache/application_info.go`, whenever there is a state transition, we can post a `UpdateResponse` to the shim. Why this cannot work?

**Comment by kingamarton** [comment]
I thing I get the point mentioned by @wilfred-s. I agree that there is no request in most of the cases, but we added this new field to the UpdateResponse, because there are handled the other Application related changes. I had an another idea to create a new object and send it back instead of creating a full UpdateRequest with only one field populated, but that way I will break the contract.

Related your comment regarding the "3 ways of communicating these updates". The other updates are for different object types, I don't this is a good idea to mix them. I would handle the application separately from the OutstandingResourceResource and contanerSchedulingState update. @wilfred-s do you have any better idea of handling this?

**Comment by wilfred-s** [comment]
We have had a long offline discussion about this and we have a problem with the way the interface was designed.
That problem cannot be fixed as part of this change. We need to work on that outside of this change. The problem with the way this is designed is that whenever we do anything we need to wrap it in an `UpdateRequest` or `UpdateResponse` message as that is the only message the interface allows us to send between the core and the shim.

I am filing a follow up jira to discuss and modify how this was designed and implemented. Having a simple contract with just one message pushes us into a complicated message handling and creation on both sides of the contract

I'll pull in my request for changes and approve the addition to the message.

**Comment by yangwwei** [comment]
Thanks @wilfred-s , I saw https://issues.apache.org/jira/browse/YUNIKORN-337 has been filed. We can discuss the refactoring later on that JIRA. This one should be good to go. Let me merge it. Thanks!

**Pull request merged**